### PR TITLE
Add one-of support via sealed classes

### PR DESCRIPTION
**Pull request by mc-lovin**

### CODE REVIEW

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 1]
```diff
@@ -144,9 +152,39 @@ class KotlinGenerator private constructor(
...
     return classBuilder.build()
   }
 
+  private fun generateOneOfClass(type: MessageType, oneOf: OneOf): TypeSpec {
+    val nameAllocator = nameAllocator(type)
+    val oneOfClassName = nameAllocator[oneOf].beginWithUpperCase()
+    val oneOfClassType = type.oneOfClass(oneOf)
+    val builder = TypeSpec.classBuilder(oneOfClassName)
+        .addModifiers(SEALED)
+    oneOf.fields().forEach {
```
Comment (oldergod) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: How about naming the parameter when the block is that long?

File: wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 2]
```diff
@@ -351,6 +389,19 @@ class KotlinGenerator private constructor(
           .build())
     }
 
+    if (!javaInterOp) {
+      message.oneOfs().forEach {
```
Comment (oldergod) [inline wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt]: name parameter please

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            OneOfMessage::class.java
+        ) {
+            override fun encodedSize(value: OneOfMessage): Int = 
+                when (value.choice) {
+                    is Choice.Foo -> ProtoAdapter.INT32.encodedSizeWithTag(1, value.choice.foo)
+                    is Choice.Bar -> ProtoAdapter.STRING.encodedSizeWithTag(3, value.choice.bar)
+                    is Choice.Baz -> ProtoAdapter.STRING.encodedSizeWithTag(4, value.choice.baz)
+                    else -> 0
+                } +
+                value.unknownFields.size
```
Comment (oldergod) [inline wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: nit: maybe that'd be easier to read if this line was first?
something like
```kotlin
value.unknownFields.size + when (value.choice) {
    is Choice.Foo -> ProtoAdapter.INT32.encodedSizeWithTag(1, value.choice.foo)
    is Choice.Bar -> ProtoAdapter.STRING.encodedSizeWithTag(3, value.choice.bar)
    is Choice.Baz -> ProtoAdapter.STRING.encodedSizeWithTag(4, value.choice.baz)
    else -> 0
}
```

Comment (oldergod) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: but on the other hand, we want unknowFields at the end

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            override fun decode(reader: ProtoReader): OneOfMessage {
+                var choice: Choice? = null
+                val unknownFields = reader.forEachTag { tag ->
+                    when (tag) {
+                        1 -> choice = Choice.Foo(ProtoAdapter.INT32.decode(reader))
+                        3 -> choice = Choice.Bar(ProtoAdapter.STRING.decode(reader))
+                        4 -> choice = Choice.Baz(ProtoAdapter.STRING.decode(reader))
+                        else -> TagHandler.UNKNOWN_TAG
+                    }
+                }
```
Comment (oldergod) [inline wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: the whole block is hard to read to me because of the side-effects mix but I don't have a better alternative right now.

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt (hunk 3 above)
Comment (Egorand) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: Plus, this is just an edge case where the proto has a single field. Formatting multiple `oneof` fields in a meaningful way would be interesting.

File: wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt (hunk 4 above)
Comment (oldergod) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: Specially I'm confused that the block assigns stuff but also returns a value

Comment (mc-lovin) [reply wire-tests/src/test/proto-kotlin/com/squareup/wire/protos/kotlin/OneOfMessage.kt]: right the way we collect unknown fields here is indeed al little hacky. I will think about this more. 

maybe instead of returning the assigned value or  `TagHandler.UNKNOWN_TAG` we can do a precheck to see if the tag belongs to the list of used tags
